### PR TITLE
Feature: Program Tracing 

### LINKED ISSUE #613: Implementation: Tracing
Implementation for #428

Replace current logging system with a tracing log(Remove all the severity levels). There

There should be 2 types of information emitted
- Diagnostics(warning or error). Those are collected and logged at the end.
- Tracing(Area based). Logged as it goes

### Low level api

Trace area is a path `.` seperated. This can be used to enable more granualar trace

Emit a trace
```ts
program.trace(area: string, message: string);
```

Configure trace

```
# emit every trace
--trace *              
# emit all trace starting with module-resolution (module-resolution.import, module-resolution.emitter, ect)
--trace module-resolution  or --trace module-resolution.*
```

### Helper api

Might be good to have a way to define the current area and a way to reduce duplication.

### PR DESCRIPTION
**Pull request by timotheeguerin**
fix #613

Create a new tracer for the program that replace the public access of the logger. Logger also removed `debug`, `verbose` and `info` from level to keep only:
- `error`: Diagnostic error
- `warn`: Diagnostic warn
- `trace`: Message sent via the tracing module

### No trace enabled **Default**

image

### Specific trace area
image

Todo: 
- [x] Docs

### CODE REVIEW

File: packages/compiler/core/logger/tracer.ts [hunk 1]
```diff
@@ -0,0 +1,109 @@
+import { Logger, Tracer, TracerOptions } from "../types.js";
```
[the remark this answers is not in the document]
Comment (nguerrera) [reply packages/compiler/core/logger/tracer.ts:1]: I like that idea.

[the remark this answers is not in the document]
Comment (bterlson) [reply packages/compiler/core/logger/tracer.ts:1]: Same

File: packages/website/src/docs/configuration/tracing.md [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+```bash
+cadl compile . --trace import-resolution
+```
+
+You can use the `--trace` option multiple times if there is multiple areas that should be logged from.
+
+```bash
+cadl compile . --trace import-resolution  --trace projection
+```
```
Comment (nguerrera) [inline packages/website/src/docs/configuration/tracing.md:20]: I take it `--trace *` is the way to get absolutely all tracing?

File: packages/website/src/docs/configuration/tracing.md [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+## Trace selection
+
+The tracing system in the cadl compiler works by having each trace under an area. The area name is a dot `.` seperate string of area segments.
+
+When filtering which area to select you can use this area path to select which area is going to be revealed.
+The filter follow the same naming style, expect the last segment could be a wildcard `*`. This is however the same result as omitting the last segement all toghether. In other words, those filter have the exact same behavior:
```
Comment (nguerrera) [inline packages/website/src/docs/configuration/tracing.md:27]: This makes me wonder why we have the `.*` syntax. Is it only at the top-level filter x != filter x.*?

Comment (nguerrera) [inline packages/website/src/docs/configuration/tracing.md:27]:
```suggestion
The filter follow the same naming style, except the last segment could be a wildcard `*`. This is however the same result as omitting the last segement all toghether. In other words, those filter have the exact same behavior:
```

File: packages/website/src/docs/configuration/tracing.md [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+You can use the `--trace` option multiple times if there is multiple areas that should be logged from.
+
+```bash
+cadl compile . --trace import-resolution  --trace projection
+```
+
+## Trace selection
+
+The tracing system in the cadl compiler works by having each trace under an area. The area name is a dot `.` seperate string of area segments.
```
Comment (nguerrera) [inline packages/website/src/docs/configuration/tracing.md:24]:
```suggestion
The tracing system in the cadl compiler works by having each trace under an area. The area name is a dot `.` separated string of area segments.
```

**Review by nguerrera — COMMENTED** [review]
Looks good overall, I had a couple of questions.

File: packages/compiler/core/cli.ts [hunk 5]
```diff
@@ -96,10 +96,15 @@ async function main() {
...
+            string: true,
+            describe: "List of areas that should have the trace shown. e.g. `import-resolution.*`",
+          })
           .option("diagnostic-level", {
             type: "string",
-            default: "info",
-            choices: ["error", "warn", "info", "verbose", "debug"],
+            default: "trace",
+            choices: ["error", "warn", "trace"],
             describe: "Diagnostics of this level or above will be reported.",
```
Comment (nguerrera) [inline packages/compiler/core/cli.ts:108]: Hmm, this is weird to me. So if I do --diagnostic-level warn, then --trace effectively does nothing? It doesn't seem like that's helpful if so. I don't think we should conflate tracing with diagnostics in the public UX.

It feels like the only real use case is to disable warnings, and maybe --nowarn or something would be more discoverable than --diagnostic-level=error.

Comment (nguerrera) [reply packages/compiler/core/cli.ts:108]: Hmm, I think --nowarn is a bad name choice for disable all warnings as coming from csc, it is used with arguments to disable specific warnings.

I wonder if we can reuse the filtering for warnings, but inverted:

`--nowarn *` Disable all warnings
`--nowarn a.b.c` Disable a.b.c.* warnings

File: packages/website/src/docs/configuration/tracing.md (hunk 2 above)
Comment (timotheeguerin) [reply packages/website/src/docs/configuration/tracing.md:20]: indeed

File: packages/website/src/docs/configuration/tracing.md (hunk 3 above)
Comment (timotheeguerin) [reply packages/website/src/docs/configuration/tracing.md:27]: I guess it is really not needed, was just thinking that;s something that feel quite intuitive to be able to do

File: packages/compiler/core/cli.ts (hunk 5 above)
Comment (timotheeguerin) [reply packages/compiler/core/cli.ts:108]: yeah, wondering if we even want this

File: packages/website/src/docs/configuration/tracing.md (hunk 2 above)
Comment (timotheeguerin) [reply packages/website/src/docs/configuration/tracing.md:20]: i'll add a doc for that

File: packages/compiler/core/cli.ts (hunk 5 above)
Comment (timotheeguerin) [reply packages/compiler/core/cli.ts]: I just removed it, I think we can revisit as a separate issue if we really want to mute warnings.

**Review by nguerrera — APPROVED** [review]

**Pull request merged**